applications/stepswitch/src/stepswitch_bridge.erl

```diff
@@ -365,120 +428,129 @@ endpoint_format_from(Endpoint, CIDNum, DefaultRealm, CCVs) ->
-    case wh_json:get_value(<<"Hunt-Account-ID">>, JObj) of
-        'undefined' -> emergency_cid_number(JObj);
-        _Else ->
-            wh_json:get_first_defined([<<"Emergency-Caller-ID-Number">>
-                                       ,<<"Outbound-Caller-ID-Number">>
-                                      ], JObj)
+-spec bridge_outbound_cid_name(wh_json:object()) -> api_binary().
+bridge_outbound_cid_name(JObj) ->
+    wh_json:get_first_defined(
+      [<<"Outbound-Caller-ID-Name">>
+       ,<<"Emergency-Caller-ID-Name">>
+      ], JObj
+     ).
+
+-spec bridge_outbound_cid_number(wh_json:object()) -> api_binary().
+bridge_outbound_cid_number(JObj) ->
+    wh_json:get_first_defined(
+      [<<"Outbound-Caller-ID-Number">>
+       ,<<"Emergency-Caller-ID-Number">>
+      ], JObj
+     ).
+
+-spec bridge_emergency_cid_name(wh_json:object()) -> api_binary().
+bridge_emergency_cid_name(JObj) ->
+    wh_json:get_first_defined(
+      [<<"Emergency-Caller-ID-Name">>
+       ,<<"Outbound-Caller-ID-Name">>
+      ], JObj
+     ).
+
+-spec bridge_emergency_cid_number(wh_json:object()) -> api_binary().
+bridge_emergency_cid_number(JObj) ->
+    wh_json:get_first_defined(
+      [<<"Emergency-Caller-ID-Number">>
+       ,<<"Outbound-Caller-ID-Number">>
+      ], JObj
+     ).
+
+-spec find_emergency_number(wh_json:object()) -> api_binary().
+find_emergency_number(JObj) ->
+    case whapps_config:get_is_true(
+           ?SS_CONFIG_CAT
+           ,<<"ensure_valid_emergency_cid">>
+           ,'false'
+          )
+    of
+        'false' ->
+            lager:debug("using first configured unverified emergency caller id"),
+            bridge_emergency_cid_number(JObj);
+        'true' -> ensure_valid_emergency_number(JObj)
     end.
 
--spec emergency_cid_number(wh_json:object()) ->
-                                  ne_binary().
-emergency_cid_number(JObj) ->
-    Account = wh_json:get_value(<<"Account-ID">>, JObj),
-    AccountDb = wh_util:format_account_id(Account, 'encoded'),
-    Candidates = [wh_json:get_ne_value(<<"Emergency-Caller-ID-Number">>, JObj)
-                  ,wh_json:get_ne_value(<<"Outbound-Caller-ID-Number">>, JObj)
-                 ],
-    Requested = wh_json:get_first_defined([<<"Emergency-Caller-ID-Number">>
-                                           ,<<"Outbound-Caller-ID-Number">>
-                                          ], JObj),
-    lager:debug("ensuring requested CID is emergency enabled: ~s", [Requested]),
+-spec ensure_valid_emergency_number(wh_json:object()) -> api_binary().
+ensure_valid_emergency_number(JObj) ->
+    AccountId = wh_json:get_value(<<"Account-ID">>, JObj),
+    lager:debug("ensuring emergency caller is valid for account ~s", [AccountId]),
+    Numbers = valid_emergency_numbers(AccountId),
+    Emergency = bridge_emergency_cid_number(JObj),
+    Outbound = bridge_outbound_cid_number(JObj),
+    case {lists:member(Emergency, Numbers), lists:member(Outbound, Numbers)} of
```

Comment:
Would you want to call a function `find_valid_emergency_numbers(Numbers, [Emergency, Outbound])` where you can recurse the 2nd param for the `lists:member/2` call? I believe we support custom caller id tags besides `external`, `internal`, and `emergency`; maybe this would allow easier ability to validate those custom caller IDs too?